common/quotas/clocked_rate_limiter.go

```diff
@@ -151,3 +162,39 @@ func (l ClockedRateLimiter) SetBurstAt(t time.Time, newBurst int) {
 func (l ClockedRateLimiter) TokensAt(t time.Time) int {
 	return int(l.rateLimiter.TokensAt(t))
 }
+
+// RecycleToken should be called when the action being rate limited was not completed
+// for some reason (i.e. a task is not dispatched because it was invalid).
+// In this case, we want to immediately unblock another process that is waiting for one token
+// so that the actual rate of completed actions is as close to the intended rate limit as possible.
+// If no process is waiting for a token when RecycleToken is called, this is a no-op.
+//
+// Since we don't know how many tokens were reserved by the process calling recycle, we will only unblock
+// new reservations that are for one token (otherwise we could recycle a 1-token-reservation and unblock
+// a 100-token-reservation). If all waiting processes are waiting for >1 tokens, this is a no-op.
+//
+// Because recycleCh is an unbuffered channel, the token will be reused for the next waiter as long
+// as there exists a waiter at the time RecycleToken is called. Usually the attempted rate is consistently
+// above or below the limit for a period of time, so if rate limiting is in effect and recycling matters,
+// most likely there will be a waiter. If the actual rate is erratically bouncing to either side of the
+// rate limit AND there are many invalid tasks, we can make it a buffered channel instead.
+//
+// Our goal is to ensure that each token in our bucket is used every second, meaning the time between
+// taking and successfully using a token must be <= 1s. For this to be true, we must have:
+//
+//	time_to_recycle * number_of_recycles_per_second <= 1s
+//	time_to_recycle * probability_of_recycle * number_of_attempts_per_second <= 1s
+//
+// Therefore, it is also possible for this strategy to be inaccurate if the delay between taking and
+// successfully using a token is greater than one second.
```

Comment:
I know we talked about this but I don't understand it from this comment. There's nothing special about 1s... where does that number come from?

It seems like the longer the delay between taking and recycling, the more rate will be "wasted", and the shorter the delay the less will be wasted. But it's a continuous thing, isn't it?

Reply:
hm, I thought that 1s comes from the fact that our rate limits are specified as RPS, so the rate limiter has a certain # of tokens to dole out each second. But I totally could have misunderstood -- I didn't read the innermost parts of the rate limit package that we import

Reply:
yeah but it doesn't add the tokens at the top of each second, it's continuous. so the units of the rate shouldn't matter.